MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -141,6 +151,71 @@ RemoveMemoryMapEntry (
   }
 }
 
+/**
+  Get the memory type for a given bucket.
+
+  @param  PhysicalStart  The starting address of the memory region.
+  @param  PhysicalEnd    The ending address of the memory region.
+
+  @return The memory type for the bucket that contains the given physical address range.
+          If the address range does not match any special bucket, it returns EfiMaxMemoryType.
+**/
+EFI_MEMORY_TYPE
+GetBucketMemoryType (
+  IN EFI_PHYSICAL_ADDRESS  PhysicalStart,
+  IN EFI_PHYSICAL_ADDRESS  PhysicalEnd
+  )
+{
+  EFI_MEMORY_TYPE  BucketType;
+
+  // Find the bucket type for the incoming memory region.
+  for (BucketType = 0; BucketType < EfiMaxMemoryType; BucketType++) {
+    //
+    // If the number of pages for this memory type is not zero, the input region
+    // better be within the same bucket.
+    //
+    if (mMemoryTypeStatistics[BucketType].Special && (mMemoryTypeStatistics[BucketType].NumberOfPages != 0)) {
+      if ((PhysicalStart >= mMemoryTypeStatistics[BucketType].BaseAddress) &&
+          (PhysicalEnd <= mMemoryTypeStatistics[BucketType].MaximumAddress))
+      {
+        break;
+      } else if (((PhysicalStart >= mMemoryTypeStatistics[BucketType].BaseAddress) &&
+                 (PhysicalStart <= mMemoryTypeStatistics[BucketType].MaximumAddress)) ||
+                 ((PhysicalEnd >= mMemoryTypeStatistics[BucketType].BaseAddress) &&
+                 (PhysicalEnd <= mMemoryTypeStatistics[BucketType].MaximumAddress)))
+      {
+        // The start and end overlap the bucket, but not fully inclusive. We should not allow this.
+        DEBUG ((
+          DEBUG_ERROR,
+          "%a: %lx-%lx intersects bucket %d (%lx-%lx)\n",

```

Comment:
To clarify the number maps to a bucket type.

```suggestion
          "%a: %lx-%lx intersects bucket type %d (%lx-%lx)\n",
```